### ISSUE TITLE
Indistinguishable actual/expected output

### ISSUE THREAD

**Issue by zaufi** [zaufi/pytest-matcher#17]
When the difference between actual/expected is e.g., only a `\n` at the end of text it's hard to find WHY the test has failed.

**Comment by xymaxim** [comment]
Yeah, I feel that. It can take almost forever to find the unseen symbol.

The diff command stupidly easy handles the issue with a message:

```
$ diff -u a b
@@ -1,2 +1,2 @@
 LINE 1
-LINE 2
\ No newline at end of file
+LINE 2
```
Let me ask you a question about an assertion message and the used actual/expected output style. I find it quite handy to look at and compare full outputs, but have you considered comparing outputs line by line before? I’m just wondering about the choice.

There’s the [pytest-clarity](https://github.com/darrenburns/pytest-clarity) plugin, which is pretty useful to show precise colorful diff, but it compares two operands line by line. Anyway, it also doesn’t handle newline characters (see https://github.com/darrenburns/pytest-clarity/issues/27).

**Comment by xymaxim** [comment]
> Let me ask you a question about an assertion message and the used actual/expected output style. I find it quite handy to look at and compare full outputs, but have you considered comparing outputs line by line before? I’m just wondering about the choice.

Nevermind, after #1 it would be possible to access the content and compare outputs in almost different ways. For example, quickly check outputs containing two visually indistinguishable glyphs:

```
    def test_example(expected_out):
        actual = "ã" # vs "ã"
>       assert actual.encode("utf-8") == str(expected_out).encode("utf-8")
        assert actual == expected_out
E       AssertionError: assert b'a\xcc\x83' == b'\xc3\xa3'
E         
E         At index 0 diff: b'a' != b'\xc3'
E         
E         Full diff:
E         - b'\xc3\xa3'
E         ?      ^  ^
E         + b'a\xcc\x83'
E         ?   +   ^  ^
```

**Comment by zaufi** [comment]
@xymaxim I've added a unified diff mode ;-)

![image](https://github.com/zaufi/pytest-matcher/assets/548715/3e895779-9a24-4b07-baee-a4a817992f6e)

**Comment by zaufi** [comment]
> Let me ask you a question about an assertion message and the used actual/expected output style. I find it quite handy to look at and compare full outputs, but have you considered comparing outputs line by line before? I’m just wondering about the choice.

It's just a *historical reason* ;-) I start this project inspired by Boost Unit Test Framework feature and similar in CMake's internal testing facilities. Yep, having diff was desirable but I just postpone implementation of this feature…

**Comment by zaufi** [comment]
With `pm-mismatch-style="full"` now it looks like this:
```
E       AssertionError: assert
E         The test output doesn't equal to the expected
E         (from `/home/exher/work/GitHub/pytest-matcher/master/test/data/expected/test_foo/test_diff.out`):
E         ---[BEGIN actual output]---
E         Hello Africa!↵
E         How are you doing?
E         ---[END actual output]---
E         ---[BEGIN expected output]---
E         Hello Nigeria!↵
E         ---[END expected output]---
```

What, I'd say, pretty clear now ;)

**Comment by xymaxim** [comment]
> It's just a _historical reason_ ;-) I've started this project inspired by Boost Unit Test Framework feature and similar in CMake's internal testing facilities. Yep, having diff was desirable but I just postpone implementation of this feature…

I'm actually used to this `full` mode by now, but it's definitely nice to have multiple modes.
